demos/guide/tls-client-block.c

```diff
@@ -0,0 +1,252 @@
+     * really know what you are doing.
+     */
+    SSL_CTX_set_verify(ctx, SSL_VERIFY_PEER, NULL);
+
+    /* Use the default trusted certificate store */
+    if (!SSL_CTX_set_default_verify_paths(ctx)) {
+        printf("Failed to set the default trusted certificate store\n");
+        goto end;
+    }
+
+    /* Create an SSL object to represent the TLS connection */
+    ssl = SSL_new(ctx);
+    if (ssl == NULL) {
+        printf("Failed to create the SSL object\n");
+        goto end;
+    }
+
+    /*
+     * Create the underlying transport socket/BIO and associate it with the
+     * connection
+     */
+    bio = create_socket_bio(HOSTNAME, PORT);
+    if (bio == NULL) {
+        printf("Failed to crete the BIO\n");
+        goto end;
+    }
+    SSL_set_bio(ssl, bio, bio);
+
+    /*
+     * Tell the server during the handshake which hostname we are attempting
+     * to connect to in case the server supports multiple hosts.
+     */
+    if (!SSL_set_tlsext_host_name(ssl, HOSTNAME)) {
+        printf("Failed to set the SNI hostname\n");
+        goto end;
+    }
+
+    /*
+     * Ensure we check during certificate verification that the server has
+     * supplied a certificate for the hostname that we were expecting.
+     * Virtually all clients should do this unless you really know what you
+     * are doing.
+     */
+    if (!SSL_set1_host(ssl, HOSTNAME)) {
+        printf("Failed to set the certificate verification hostname");
+        goto end;
+    }
+
+    /* Do the handshake with the server */
+    if (SSL_connect(ssl) < 1) {
+        printf("Failed to connect to the server\n");
+        /*
+         * If the failure is due to a verification error we can get more
+         * information about it from SSL_get_verify_result().
+         */
+        if (SSL_get_verify_result(ssl) != X509_V_OK)
+            printf("Verify error: %s\n",
+                X509_verify_cert_error_string(SSL_get_verify_result(ssl)));
+        goto end;
+    }
+
+    /* Write an HTTP GET request to the peer */
+    if (!SSL_write_ex(ssl, request, strlen(request), &written)) {
+        printf("Failed to write HTTP request\n");
+        goto end;
+    }
+
+    /*
+     * Get up to sizeof(buf) -1  bytes of the response. OpenSSL does not
+     * guarantee that the data is a string or is NUL terminated. We reserve
+     * the last byte in the buffer for a trailing NUL which we add
+     * afterwards. We keep reading until the server closes the connection.
+     */
+    while (SSL_read_ex(ssl, buf, sizeof(buf) - 1, &readbytes)) {
+        /*
+        * We add a trailing NUL terminator and treat the response as a
+        * string for the purposes of this simple example. In fact OpenSSL
+        * does not guarantee this and in reality the response might be
+        * non-printable or have NUL characters in the middle of it.
+        */
+        buf[readbytes] = '\0';
+        printf("%s", buf);
```

Comment:
Could just use fwrite(3) to dump `readbytes` bytes of the response, and then putchar('\n').

Reply:
Also, what if the server does not close the connection as soon as the response is sent? Perhaps also include a "Connection: close\r\n" request header?

Reply:
I'm doing HTTP/1.0 here. IIUC, in HTTP/1.0 the default expectation is that the connection is closed automatically. The "Connection: close" header is an HTTP/1.1 header so shouldn't be needed. Or have I misunderstood something?